go/merkletree2/config.go

```diff
@@ -19,28 +30,34 @@ type Config struct {
 	// simplicity, we enforce that all the keys have the same length and that
 	// bitsPerIndex divides keyByteLength*8
 	keysByteLength int
-
-	// Construct a new object to unmarshal values into
-	valueConstructor ValueConstructor
 }
 
-// NewConfig makes a new config object. It takes a a Hasher (for example
-// sha512.Sum512), logChildrenPerNode which is the base 2 logarithm of the
-// number of children per interior node, valuesPerLeaf the maximum number of
-// entries in a leaf before the leaf is split into multiple nodes (at a lower
-// level in the tree), keyByteLength the length of the Keys which the tree
-// will store, and a valueConstructor (so that typed values can be pulled out
-// of the Merkle Tree).
-func NewConfig(h Hasher, logChildrenPerNode uint8, valuesPerLeaf int, keysByteLength int, valueConstructor ValueConstructor) (Config, error) {
+// NewConfig makes a new config object. It takes a a Hasher, logChildrenPerNode
+// which is the base 2 logarithm of the number of children per interior node,
+// valuesPerLeaf the maximum number of entries in a leaf before the leaf is
+// split into multiple nodes (at a lower level in the tree), keyByteLength the
+// length of the Keys which the tree will store, and a valueConstructor (so that
+// typed values can be pulled out of the Merkle Tree).
+func NewConfig(h Hasher, useBlindedValueHashes bool, logChildrenPerNode uint8, valuesPerLeaf int, keysByteLength int) (TreeConfig, error) {
 	childrenPerNode := 1 << logChildrenPerNode
 	if (keysByteLength*8)%int(logChildrenPerNode) != 0 {
-		return Config{}, NewInvalidConfigError("The key bit length does not divide logChildrenPerNode")
+		return TreeConfig{}, NewInvalidConfigError("The key bit length does not divide logChildrenPerNode")
+	}
+	if logChildrenPerNode > 63 {
+		return TreeConfig{}, NewInvalidConfigError("This package does not support more than 2^63 children per internal node")
 	}
-	return Config{hasher: h, childrenPerNode: childrenPerNode, valuesPerLeaf: valuesPerLeaf, bitsPerIndex: logChildrenPerNode, keysByteLength: keysByteLength, valueConstructor: valueConstructor}, nil
+	return TreeConfig{hasher: h, useBlindedValueHashes: useBlindedValueHashes, childrenPerNode: childrenPerNode, maxValuesPerLeaf: valuesPerLeaf, bitsPerIndex: logChildrenPerNode, keysByteLength: keysByteLength}, nil
 }
 
+type MasterSecret []byte
+type KeySpecificSecret []byte
+
 // Hasher is an interface for hashing MerkleTree data structures into their
-// cryptographic hashes.
+// cryptographic hashes. It also manages blinding secrets.
 type Hasher interface {
-	Hash([]byte) Hash
+	EncodeAndHashGeneric(interface{}) (Hash, error)
+	HashKeyValuePairWithMasterSecret(KeyValuePair, Seqno, MasterSecret) (Hash, error)
+	HashKeyValuePairWithKeySpecificSecret(KeyValuePair, KeySpecificSecret) (Hash, error)
+	GenerateMasterSecret(Seqno) (MasterSecret, error)
```

Comment:
I was trying to keep it more general, in case we want to derive a MasterSecret in a pseudorandom way (for example by hashing a fixed secret with the seqno). Can take it out though.

Reply:
Yes.